Set up email functionality for moderators to use when unpublishing or featuring explorations.

seanlip opened the pull request:
This PR implements functionality for sending emails after unpublishing or featuring an exploration. It follows a spec that @jacobdavis11 and I worked together on in Aug 2015.

@jacobdavis11 -- would you mind taking a fairly careful look at this (code, functionality and UI)? Please note that, when testing it on your local machine, you will need to set `feconf.CAN_SEND_EMAILS_TO_USERS` and `feconf.REQUIRE_EMAIL_ON_MODERATOR_ACTION` to True.

Thanks!

seanlip commented:
@jacobdavis11 -- one question, who do you think the resulting email should be "from"? Should it be from the moderator's username (using the system email address), or should it be from the system username and email address and contain a line similar to "... has sent you a message" at the top?

jacobdavis11 commented on core/controllers/moderator.py:
```diff
@@ -49,3 +50,15 @@ def post(self):
         })
+
+
+class EmailDraftHandler(base.BaseHandler):
```
It seems strange that this is supplying the content but not the subject; I feel the two should be treated equivalently throughout the code.

seanlip replied:
My intention here is that only the content is editable.

jacobdavis11 commented on core/domain/email_manager.py:
```diff
@@ -63,31 +65,43 @@
-# NOTE TO DEVELOPERS: post-signup emails will not be sent if this placeholder
-# is left unmodified. If this policy changes, this should be documented in the
-# wiki.
+# NOTE TO DEVELOPERS: the relevant emails will not be sent if any of these
```
This looks like I'm supposed to modify these strings directly in this file, but I can also modify them via the admin config so I am confused.

seanlip replied:
Yep, this is misleading. I removed it (also in view of your general comment 1).

jacobdavis11 commented on core/domain/email_manager.py:
```diff
@@ -164,5 +178,71 @@ def send_post_signup_email(user_id):
+        raise Exception('Unrecognized email intent: %s' % intent)
+
+
+def _get_email_config(intent):
```
I don't see a difference between an 'action' and an 'intent' and you've already checked that the action is valid in controllers, so I think this can be omitted.

seanlip replied:
It actually took me a while to hit on this idea, because somewhere along the line I needed to change 'user action' into 'intent of email being sent' -- and this is what this represents. I think this is a good solution, and that the intent/action string being the same for the existing actions is only coincidental. They really represent different things conceptually, so I think that having this seam is fine.

jacobdavis11 replied:
Could you give an example where the action is not the same as the intent?

seanlip replied:
Technically, this is the case in these examples: the intent is a subset of the action. What I mean by that is, the action of unpublishing an exploration consists of two things: changing the exploration's status, and sending an email. The intent refers only to the email. Perhaps it might not have been a good idea to use the same constant strings for these things, but I think it's really the names of the constants that matter so I'm not _too_ worried.

You could imagine a future scenario where moderators had a web form that allowed them to essentially send a message to any contributor. They would be able to choose from a list of intents (which makes searching past emails easier in the future), and compose an arbitrary email body. The action here is basically 'send private message'; the intent is whatever the moderator picks.

jacobdavis11 commented on core/domain/email_manager.py:
```diff
@@ -164,5 +178,71 @@ def send_post_signup_email(user_id):
+
+    If feconf.REQUIRE_EMAIL_ON_MODERATOR_ACTION is set to False, returns None.
+    """
+    if not feconf.REQUIRE_EMAIL_ON_MODERATOR_ACTION:
```
You've already checked this in controllers. I think this file is justified in assuming that everything is legitimate and it can just send the emails.

seanlip replied:
Agreed. Removed.

jacobdavis11 commented on core/domain/email_manager.py:
```diff
@@ -164,5 +178,71 @@ def send_post_signup_email(user_id):
+    sender_user_settings = user_services.get_user_settings(sender_id)
+    email_subject = feconf.VALID_MODERATOR_ACTIONS[intent]['email_subject_fn'](
+        exploration_title)
+    full_email_content = (
```
I think the Hi and Thanks parts should be included in the message sent by the frontend, with {{}} placeholders for names. Basically I'm concerned that too much fungible stuff is being hard-coded here.

seanlip replied:
I added a message in the frontend to tell them not to do that, and to explain how the email message is constructed. I don't want to implement placeholders and interpolation at the moment; this will take a while and it's possible to get it wrong.

jacobdavis11 replied:
If you want to do things this way then I suggest storing the salutation and sign-off lines as functions in feconf, in the same way as the subject lines. That way the emails will only be in two different places (feconf and the admin configs).

seanlip replied:
That seems sensible. Done.

jacobdavis11 commented on core/templates/dev/head/editor/ExplorationSettings.js:
```diff
@@ -247,16 +247,66 @@ oppia.controller('ExplorationSettings', [
+    warningsData.clear();
+
+    $http.get('/moderatorhandler/email_draft/' + action).then(function(response) {
+      // An empty draft email body indicates that no email will be sent.
```
This isn't communicated to the moderator in the UI.

seanlip replied:
This was unclear; the comment meant that if the email draft was empty, the email form wouldn't be shown in the UI at all. I clarified it.

jacobdavis11 commented:
The UI looks good. In reply to your question I would suggest a system address of some kind so that people don't think they can reply to the email. A few comments:

1. Is the requirement to set emails contents on the config tab specified in the setup instructions? The user may not notice the error messages that the backend is throwing and just think there's an error (I did at first) especially as the warning sent to the front end is a generic "Error communicating with server". Also as the moderator gets to read and edit the email before sending it I'm not sure forcing people to use the admin config is necessary here; you could just include a generic "this exploration has been unpublished" by default.

2. There don't seem to be options to set subject lines of the emails on the admin page. If you want it to be hard-coded then I suggest something more Oppia-specific like "Your Oppia exploration 'how to count' has been unpublished".

3. It's a bit weird that the unpublishing and featuring options aren't next to each other in the admin config list (the gallery carousel comes between them).

4. You may want to offer individual choices as to when to send emails; for example someone might want to send them when unpublishing but not when featuring - in fact we might want to do that. These options could be either in feconf or in the admin config, probably the latter is better.

5. The email editor allows all rich text components, including collapsibles and picture upload. I'm not sure what happens to the pictures and didn't get any errors, but I can't see the emails that are being sent and suspect this will end badly.

seanlip commented on core/controllers/editor.py:
```diff
@@ -481,6 +482,68 @@ def put(self, exploration_id):
+
+    @base.require_moderator
+    def put(self, exploration_id):
+        """Updates the editing rights for the given exploration, and sends an
```
Good catch! Fixed.

seanlip commented on core/controllers/editor.py:
```diff
@@ -481,6 +482,68 @@ def put(self, exploration_id):
+        # If moderator emails can be sent, check that all the prerequisites are
+        # satisfied, otherwise do nothing.
+        if feconf.REQUIRE_EMAIL_ON_MODERATOR_ACTION:
+            if not email_body:
```
This was an error, thanks for catching it. I used a HTML field when it should have been a unicode field (sometimes the former inserts stuff like '<p></p>').

seanlip commented:
Thanks, @jacobdavis11 -- I pushed another commit. PTAL!

Regarding your general comments:

1. I removed the need to use the admin config. The original implementation was too complicated, so thanks for calling that out.

2. I do want these to be hard-coded; I changed the subject lines per your suggestion.

3. The admin page is a mess at the moment. The config options are essentially ordered at random; I don't think I'll fix this in this commit.

4. I'll punt on this for now -- maybe we'll do it in the future.

5. Good catch; this was due to the error I alluded to in one of my commits. It's been fixed.

jacobdavis11 commented on core/domain/email_manager.py:
```diff
@@ -164,5 +181,60 @@ def send_post_signup_email(user_id):
+
+def _get_email_config(intent):
+    require_valid_intent(intent)
+    return config_domain.Registry.get_config_property(
```
I tried chasing this through config_domain but I couldn't find where the default email texts are now stored. Can you enlighten me?

seanlip replied:
It's in the definition of the config properties in this file -- search for PUBLICIZE_EXPLORATION_EMAIL_HTML_BODY. It's the last argument.

jacobdavis11 replied:
That doesn't seem like a message you could actually send to a user. Couldn't you make it some generic message about what has happened so that editing it is optional?

Also could you move this variable to feconf? Then all the email-related texts will be in the same place, and people are more likely to notice and edit them.

seanlip replied:
Good point, I changed the message. The help text in the editor UI should be sufficient.

However, I'm leaving the message where it is. These config options should only be edited in feconf or in the admin panel, but not both, and none of our other config properties declare their default values in feconf. In any case, this is just the default value for the config option, and app owners can set it to whatever they wish in the admin panel.

seanlip commented:
Thanks @jacobdavis11, I made another commit fixing the issues. PTAL.
